### PR TITLE
bpo-31658: Make xml.sax.parse accepting Path objects

### PR DESCRIPTION
**Pull request by BoboTiG**
https://bugs.python.org/issue31658

### CODE REVIEW

File: Lib/xml/sax/saxutils.py [hunk 1]
```diff
@@ -340,7 +341,7 @@ def prepare_input_source(source, base=""):
     """This function takes an InputSource and an optional base URL and
     returns a fully resolved InputSource object ready for reading."""
 
-    if isinstance(source, str):
+    if isinstance(source, (str, pathlib.Path)):
```
Comment (merwok) [inline Lib/xml/sax/saxutils.py]: What about using [os.fspath](https://docs.python.org/3/library/os.html#os.fspath) for duck-typing here?

Comment (BoboTiG) [reply Lib/xml/sax/saxutils.py]: Something like that?
```python
if  isinstance(source, str) or hasattr(source, "__fspath__"):
```

Comment (merwok) [reply Lib/xml/sax/saxutils.py]: More like:

```python
if isinstance(source, os.PathLike):
    source = os.fspath(source)
```

Comment (BoboTiG) [reply Lib/xml/sax/saxutils.py]: Done, it is much better now, thanks :)

File: Doc/library/xml.sax.reader.rst [hunk 2]
```diff
@@ -102,13 +102,16 @@ The :class:`XMLReader` interface supports the following methods:
 
    Process an input source, producing SAX events. The *source* object can be a
    system identifier (a string identifying the input source -- typically a file
-   name or a URL), a file-like object, or an :class:`InputSource` object. When
-   :meth:`parse` returns, the input is completely processed, and the parser object
-   can be discarded or reset.
+   name or a URL), a :class:`~pathlib.Path` object, a file-like object or an
```
Comment (merwok) [inline Doc/library/xml.sax.reader.rst]: Is there a glossary entry for `file-like` that could be linked here?

Comment (BoboTiG) [reply Doc/library/xml.sax.reader.rst]: I found those usages:
```shell
$ rgrep -r '`file-like' . | grep rst | grep file-like
./Doc/glossary.rst:      etc.).  File objects are also called :dfn:`file-like objects` or
./Doc/library/importlib.rst:        Returns an opened, :term:`file-like object` for binary reading
./Doc/library/xml.sax.reader.rst:   name or a URL), a :class:`~pathlib.Path` object, a :term:`file-like object`
./Doc/library/email.generator.rst:   to the :meth:`write` method, to the :term:`file-like object` *outfp*.
./Doc/library/email.generator.rst:   method, to the :term:`file-like object` *outfp*.  *outfp* must support a
./Doc/library/json.rst:   :term:`file-like object`) using this :ref:`conversion table
./Doc/library/tempfile.rst:   Return a :term:`file-like object` that can be used as a temporary storage area.
./Doc/library/asyncio-eventloop.rst:   *pipe* is a :term:`file-like object <file object>`.
./Doc/library/asyncio-eventloop.rst:   *pipe* is :term:`file-like object <file object>`.
```
So I selected the most used:
```rst
    :term:`file-like object`
```

File: Lib/xml/sax/saxutils.py (hunk 1 above)
Comment (merwok) [reply Lib/xml/sax/saxutils.py]: I remembered the name wrong (path-like), so I fixed it directly

**Comment by BoboTiG** [comment]
@merwok do you think we can move on?

File: Lib/test/test_sax.py [hunk 3]
```diff
@@ -397,6 +402,13 @@ def test_string(self):
         self.checkContent(prep.getByteStream(),
                           b"This was read from a file.")
 
+    def test_path_objects(self):
+        # If the source is a Path object, use it as a system ID and open it.
+        prep = prepare_input_source(pathlib.Path(self.file))
```
Comment (serhiy-storchaka) [inline Lib/test/test_sax.py]: Use `test.support.FakePath` instead of `pathlib.Path` for testing.

Comment (BoboTiG) [reply Lib/test/test_sax.py]: Done :)

File: Misc/NEWS.d/next/Library/2018-07-30-12-00-15.bpo-31658._bx7a_.rst [hunk 4]
```diff
@@ -0,0 +1 @@
+Make xml.sax.parse accepting Path objects. Patch by Mickaël Schoentgen.
```
Comment (tirkarthi) [inline Misc/NEWS.d/next/Library/2018-07-30-12-00-15.bpo-31658._bx7a_.rst]: Markup string can be used for linking. How about the below


:func:`xml.sax.parse` now supports :term:`path-like object`.
Patch by Mickaël Schoentgen.

Comment (merwok) [reply Misc/NEWS.d/next/Library/2018-07-30-12-00-15.bpo-31658._bx7a_.rst]: Could you double-check that this is true for NEWS entries?

Comment (tirkarthi) [reply Misc/NEWS.d/next/Library/2018-07-30-12-00-15.bpo-31658._bx7a_.rst]: @merwok Yes, just verified the links with a local build. Screenshot 

screen shot 2019-01-08 at 9 48 06 pm

Comment (BoboTiG) [reply Misc/NEWS.d/next/Library/2018-07-30-12-00-15.bpo-31658._bx7a_.rst]: Done :+1:

**Comment by BoboTiG** [comment]
What is the status? Do you think it needs more work?

**Comment by merwok** [comment]
Code looks good to me, but I haven’t seen XML in years so can’t take responsibility for merging.

**Comment by merwok** [comment]
Thanks again for the patch and thanks @scoder for merging!

**Pull request merged**